platforms/jvm/plugins-java-base/src/main/java/org/gradle/api/plugins/internal/DefaultJavaFeatureSpec.java

```diff
@@ -96,28 +105,28 @@ private void setupConfigurations(SourceSet sourceSet) {
         JvmFeatureInternal feature = new DefaultJvmFeature(name, sourceSet, capabilities, project, true, SourceSet.isMain(sourceSet));
         feature.withApi();
 
-        // TODO: #23495 Investigate the implications of using this class without
-        //       the java plugin applied, and thus no java component present.
-        //       In the long run, all domain objects created by this feature should be
-        //       owned by a component. If we do not add them to the default java component,
-        //       we should be adding them to a user-provided or new component instead.
-        project.getComponents().withType(JvmSoftwareComponentInternal.class).configureEach(component -> {
-            AdhocComponentWithVariants adhocComponent = (AdhocComponentWithVariants) component;
-            if (withJavadocJar) {
-                feature.withJavadocJar();
-                Configuration javadocElements = feature.getJavadocElementsConfiguration();
-                adhocComponent.addVariantsFromConfiguration(javadocElements, new JavaConfigurationVariantMapping("runtime", true));
-            }
-            if (withSourcesJar) {
-                feature.withSourcesJar();
-                Configuration sourcesElements = feature.getSourcesElementsConfiguration();
-                adhocComponent.addVariantsFromConfiguration(sourcesElements, new JavaConfigurationVariantMapping("runtime", true));
-            }
-
-            if (allowPublication) {
-                adhocComponent.addVariantsFromConfiguration(feature.getApiElementsConfiguration(), new JavaConfigurationVariantMapping("compile", true));
-                adhocComponent.addVariantsFromConfiguration(feature.getRuntimeElementsConfiguration(), new JavaConfigurationVariantMapping("runtime", true));
-            }
-        });
+        return feature;
+    }
+
+    private void addFeatureToComponent(JvmFeatureInternal feature, JvmSoftwareComponentInternal component) {
```

Comment:
🤔 Shouldn't this method live inside the component?

Reply:
I moved this logic up to the `DefaultJavaPluginExtension`. The primary reason this cannot be moved to the component (at least currently) is since the values for the `JavaConfigurationVariantMapping` instances are specific per feature. Features created by `registerFeature` have one mapping, while the main feature has a different mapping. 

That means this logic is not generic and thus does not belong in the component. Hopefully this situation can be improved in the future with an improved variant API.